### ISSUE TITLE
Updating to newer clang version / merging with SPIRV-LLVM

### ISSUE THREAD

**Issue by pierremoreau** [KhronosGroup/SPIR#4]
Hello,

What is the current status for this repository? Are there any plan to port the code to a newer version of clang, which would make it possible to use clang with SPIR support alongside LLVM with SPIR-V support, without having to use two different versions of LLVM? Or maybe, even try to upstream SPIR support?

**Comment by dneto0** [comment]
I don't know of any plans to do so.

The difficulty with upgrading is that SPIR 1.x is a tweak on LLVM 3.2 and SPIR 2.0 provisional is a tweak  on LLVM 3.4.  If moving to a new version of LLVM the codebase would have to add a bitcode writer for the older target LLVM version for that SPIR version.  It's non-trivial work to do so, and has not yet been discussed with the LLVM community as far as I know.

**Comment by pierremoreau** [comment]
SPIR 1.x is based on LLVM 3.2, SPIR 2.0 on LLVM 3.4, SPIR-V 1.0 backend is written on top of 3.6, so I guess SPIR-V 2.0 backend will be on top of 3.8 and none of them will be upstreamed by then? :-p

My question was more targeted towards SPIR-V, as `llvm-spirv` uses SPIR friendly LLVM code as an entry. And since there has been some discussion on upstreaming the LLVM <=> SPIR-V backend, I guess the current dependency on SPIR and an old LLVM version has been discussed?

**Comment by dneto0** [comment]
I believe the plan is that SPIR-V support is being upstreamed as soon as possible.  :-)

One of the good things about SPIR-V is that it's different enough from LLVM that it's easy to think of SPIR-V support as just another LLVM backend. (And another front end, to convert from SPIR-V to LLVM.)

**Comment by pierremoreau** [comment]
That's what I've heard as well (for both points). But then, how is it supposed to work since `llvm-spirv` will only accept an input file produced by `clang -cc1 -emit-llvm-bc -triple <triple> <OpenCL compile options> -cl-spir-compile-options "<OpenCL compile options>" -include <opencl_spir.h> -o <output> <input>`? (Or by calling libSPIRV directly from some source code?) Or did I misunderstood something?

**Comment by dneto0** [comment]
I have no idea.  I have not been paying that close attention.
@yxsamliu has been doing most of the work.

**Comment by pierremoreau** [comment]
Do you know where I could find the public discussion threads around the backend? I haven't seen any discussion on the llvm-dev ML recently, was it moved somewhere else or made private for the moment being?

**Comment by yxsamliu** [comment]
Last time I proposed to upstream the LLVM/SPIR-V converter to the community, the community would like to have a conventional backend using the SelectionDAG/MC approach. The current LLVM/SPIR-V converter is not ready for upstreaming since it does not use SelectionDAG/MC approach.

However, we are in a better position for developing a SPIR-V backend than before since we've factored out most of the OpenCL specific transformations and defined the SPIR-V friendly LLVM format. The challenge is that SPIR-V is a high level IR which is roughly at the same level as LLVM IR. It does not define registers nor stack frames and is not easy to fit into the LLVM codegen framework.

I did some prototyping for the SPIR-V backend but currently I am busy with other work, so it may take some time to have a functional prototype.

**Comment by pierremoreau** [comment]
Thank you for the explanation! I don't have any knowledge about LLVM internals nor the time to help on it right now, but do you have a branch somewhere with your current WIP which I could have a look at, try to understand what is going and then try to help?

By the way, is there a preferred way for asking these kind of questions? Opening an issue just for that seems weird.

**Comment by yxsamliu** [comment]
I can create a branch spirv-backend-prototype for that.

I think it is OK to create issues for such discussions. I don't know a better way to do that.

**Comment by pierremoreau** [comment]
That would be great, thanks!

**Comment by bader** [comment]
If I understand this request correctly it's about upgrading clang to 3.6 version. I'm going to create a separate branch in clang repository and push OpenCL 1.2/2.0 front-end compiler, which will produce SPIR-V friendly LLVM or even SPIR-V directly. The main motivation is that old SPIR formats doesn't preserve some information required for SPIR-V, so we are have to create a separate compiler for SPIR-V. If it's based on the same LLVM version as SPIR-V translator we could build clang with spirv library or spirv-backend to get single tool producing SPIR-V from OpenCL.

**Comment by pierremoreau** [comment]
To give some context to my request: I'm working on converting SPIR-V to the IR used by Nouveau (the open-source driver for NVIDIA card) as a way to run OpenCL kernels on that driver. clover (the part of Mesa dealing with OpenCL) uses LLVM and clang to compile the OpenCL code to LLVM IR, which is then parsed by the Radeon driver. I would like to do something similar, but by compiling to SPIR-V instead. Currently I need a 3.2 LLVM setup to generate the SPIR friendly LLVM code, a 3.6 LLVM install to convert the LLVM code to SPIR-V, and a 3.8 LLVM install for clover (part of the Mesa project handling the OpenCL API implementation). I should be able to skip the 3.8 install and use the 3.6 one directly but I will still need the 3.2 one, and I have to do some steps manually. (I haven't tried to call methods from both the 3.6 and the 3.2 version, but I doubt I could bind to two different versions of a same library.)
So to sum up, the main point of my request was to upgrade to the same version as the one used by SPIRV-LLVM to have a single tool producing SPIR-V from OpenCL.

Thank you for the branch! I'll make sure to test it and help improve it.

**Comment by yxsamliu** [comment]
I have created branch spirv-backend for SPIRV target. Currently most SPIRV enum values have been added. However it still needs lots of work to generate SPIRV.

**Comment by pierremoreau** [comment]
Awesome! Cloning that right now. I'll dive in LLVM and the backend code to try to help.
Closing as the backend branch solves the _issue_.
